hive/replays/recurrent_replay.py

```diff
@@ -0,0 +1,243 @@
+            stack_size=1,
+            n_step=n_step,
+            gamma=gamma,
+            observation_shape=observation_shape,
+            observation_dtype=observation_dtype,
+            action_shape=action_shape,
+            action_dtype=action_dtype,
+            reward_shape=reward_shape,
+            reward_dtype=reward_dtype,
+            extra_storage_types=extra_storage_types,
+            num_players_sharing_buffer=num_players_sharing_buffer,
+        )
+        self._max_seq_len = max_seq_len
+
+    def add(self, observation, action, reward, done, **kwargs):
+        """Adds a transition to the buffer.
+        The required components of a transition are given as positional arguments. The
+        user can pass additional components to store in the buffer as kwargs as long as
+        they were defined in the specification in the constructor.
+        """
+
+        if self._episode_start:
+            self._pad_buffer(self._max_seq_len - 1)
+            self._episode_start = False
+        transition = {
+            "observation": observation,
+            "action": action,
+            "reward": reward,
+            "done": done,
+        }
+        transition.update(kwargs)
+        for key in self._specs:
+            obj_type = (
+                transition[key].dtype
+                if hasattr(transition[key], "dtype")
+                else type(transition[key])
+            )
+            if not np.can_cast(obj_type, self._specs[key][0], casting="same_kind"):
+                raise ValueError(
+                    f"Key {key} has wrong dtype. Expected {self._specs[key][0]},"
+                    f"received {type(transition[key])}."
+                )
+        if self._num_players_sharing_buffer is None:
+            self._add_transition(**transition)
+        else:
+            self._episode_storage[kwargs["agent_id"]].append(transition)
+            if done:
+                for transition in self._episode_storage[kwargs["agent_id"]]:
+                    self._add_transition(**transition)
+                self._episode_storage[kwargs["agent_id"]] = []
+
+        if done:
+            self._episode_start = True
+
+    def _get_from_array(self, array, indices, num_to_access=1):
+        """Retrieves consecutive elements in the array, wrapping around if necessary.
+        If more than 1 element is being accessed, the elements are concatenated along
+        the first dimension.
+        Args:
+            array: array to access from
+            indices: starts of ranges to access from
+            num_to_access: how many consecutive elements to access
+        """
+        full_indices = np.indices((indices.shape[0], num_to_access))[1]
+        full_indices = (full_indices + np.expand_dims(indices, axis=1)) % (
+            self.size() + self._max_seq_len + self._n_step - 1
+        )
+        elements = array[full_indices]
+        elements = elements.reshape(indices.shape[0], -1, *elements.shape[2:])
```

Comment:
Functions and code resued from CircularReplayBuffer? Maybe just use the inherited functions? Or are there changes across all the functions?

Reply:
We are using the inherited functions defined in CircularReplayBuffer except for the ones where `max_seq_len` and `stack_size` differ.

Reply:
Can something like super calls be done? For e.g. the sample function in PPOReplayBuffer? That is the input is updated such that it can be passed to the function of the master class.

Reply:
There will be a new PR that implements a recurrent replay that saves trajectories rather than sequences of transitions. It will be more suitable for recurrent DQN. For now I think it's fine to keep those functions where `max_seq_len` and `stack_size` are differerent.